### ISSUE TITLE
OMEdit always opens ExternalMedia 4.0.0 even if I ask for ExternalMedia 3.3.2

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#10595]
We recently added ExternalMedia to the Package Manager, versions 3.3.2 (uses MSL 3.2.3) and 4.0.0 (uses MSL 4.0.0). 

I installed both on my PC using OMEdit. However, if I select version 3.3.2 from the SystemLibrary button in the Welcome screen, version 4.0.0 is loaded.

I checked the `.openmodelica` directory and indeed both versions are there. Also the `json` metadata looks good to me.

@sjoelund, @adeas31 would you mind having a look?

**Comment by adeas31** [comment]
Did you check what API call is done in the log? If the log says `loadModel(ExternalMedia, "3.3.2")` and still version `4.0.0` is loaded then the compiler is doing something wrong.

**Comment by sjoelund** [comment]
It does this for me on command-line as well. Seems really odd, but I don't have time to go into details

**Comment by perost** [comment]
ExternalMedia 4.0.0 has a conversion-annotation that states that it's backwards compatible without conversion with 3.3.2, 3.3.1 and 3.3.0, and the compiler will load the latest compatible version unless `requireExactVersion` is set.

**Comment by adeas31** [comment]
OMEdit does not use `requireExactVersion`. I will set `requireExactVersion` to `true` for both `loadModel` and `loadFile` APIs in OMEdit.

**Comment by perost** [comment]
> OMEdit does not use `requireExactVersion`. I will set `requireExactVersion` to `true` for both `loadModel` and `loadFile` APIs in OMEdit.

No, we don't want that. We want to load the latest compatible version, as per @casella own request. The issue is that the library claims to be backwards compatible even when it's not.

**Comment by casella** [comment]
@perost is right, we shouldn't use requireExactVersion by default. If I load a library that has `uses(Modelica(version=3.2.2))`, we shouldn't load that version, but rather load MSL 3.2.3, which is backwards compatible with the one declared in the uses annotation.

**Comment by casella** [comment]
Regarding ExternalMedia, the conversion annotations declare that there is no need to run any ExternalMedia-specific conversion script to update a model from using ExternalMedia 3.3.2 to using ExternalMedia 4.0.0. However, as discussed extensively in modelica/ModelicaSpecification#3209, this doesn't mean that ExternalMedia 4.0.0 is backwards-compatible. In fact, it isn't, because it uses MSL 4.0.0, which is not backwards-compatible with MSL 3.2.3, which is used by ExternalMedia 3.3.2.

The point is, if I have a model that uses ExternalMedia 3.3.2 (and thus MSL 3.2.3!), I can't load ExternalMedia 4.0.0 instead, even though that would not require running any ExternalMedia conversion scripts. The reason is that ExternalMedia 4.0.0 uses MSL 4.0.0, so I should load that version of MSL, and then convert my model from MSL 3.3.2 to MSL 4.0.0. 

That would generate a new version of my model, which is forever bound to MSL 4.0.0. That's not really "backwards compatible". Backwards compatible means I can use a new library without any change to my source code, which is not the case here.

Hence, ExternalMedia 4.0.0 is _not_ backwards-compatible with ExternalMedia 3.2.3.

**Comment by casella** [comment]
As I understand, the problem is that the current package manager generates backwards-compatibility information only by looking at the `noneFromVersion` annotations. That is not enough, it should also check if there is a relevant `noneFromVersion` annotation _in the corresponding dependencies_.

For example, ExternalMedia 4.0.0 contains the following annotation:
```modelica
  annotation(
             uses(Modelica(version="4.0.0")),
             version="4.0.0",
             conversion(
                        noneFromVersion="3.3.2",
                        noneFromVersion="3.3.1",
                        noneFromVersion="3.3.0"),
   ...
```
and the package manager script generates this code fragment in `rawdata.json`:
```json
      "4.0.0": {
        "libs": {
          "ExternalMedia": {
            "path": "ExternalMedia",
            "provides": [
              "3.3.0",
              "3.3.1",
              "3.3.2"
            ],
            "uses": {
              "Modelica": "4.0.0"
            },
            "version": "4.0.0"
          }
```
where each `noneFromVersion` string is mapped directly into an entry of the `provides` property. As the ExternalMedia case demonstrates, this is not good enough. We should instead run the following algorithm:

- Let `ver` be the version of the library, as specified by the `version` annotation. In the specific example `ver={"ExternalMedia", "4.0.0"}`
- Fetch the list of dependencies `ver_dep_list` from the uses annotation of the `ver` library. In this example  `ver_dep_list = {{"Modelica", "4.0.0"}}`
- Build an initial list of _candidate_ backwards-compatible versions `ver_back_list` with one element for each `noneFromVersion` annotations of `ver`. In this specific example `ver_back_list = {"3.3.2", "3.3.1", "3.3.0"}`
- For each candidate `ver_back` in `ver_back_list`, fetch the list of dependencies `ver_back_dep_list` from the contents of its `uses` annotation. For example, when considering `ver_back = "3.3.2"`, `ver_back_dep_list = {{"Modelica", "3.2.3"}}`
- For each package `ver_back_dep` in `ver_back_dep_list`, find the corresponding package `dep` in `ver_dep_list`. In this case, `ver_back_dep = {"Modelica", "3.2.3"}`, and the corresponding package is `dep = {"Modelica", "4.0.0"}`
- Check if the version of `ver_back_dep` is the same of `dep` (it isn't, in the example) or if the version of `ver_back_dep` shows up in the `noneFromVersion`  annotation of `dep` (it doesn't, in this case). If not, then remove the candidate backwards-compatible version  `ver_back` from `ver_back_list`
- When finished with the foreach loops, dump the surviving versions within `ver_back_list` into the "provides" property of the JSON file
 
@sjoelund, does this make sense to you?
